src/Lender.sol

```diff
@@ -1090,6 +526,52 @@ contract Lender {
         premiums[u][m] = 0;
     }
 
+    /// @notice Allows users to lend underlying asset for Illuminate PTs via an approved protocol
+    /// @param p principal value according to the MarketPlace's Principals Enum
+    /// @param u underlying asset address of the market's tuple
+    /// @param m timestamp of maturity of the market's tuple
+    /// @param a amount in underlying to be lent
+    /// @param d data to conduct the call
+    function lend(
+        uint8 p,
+        address u,
+        uint256 m,
+        uint256 a,
+        bytes calldata d
+    ) external returns (uint256) {
+        // Receive the funds from the user
+        Safe.transferFrom(IERC20(u), msg.sender, address(this), a);
+
+        // Fetch the adapter for this lend call
+        address adapter = IMarketPlace(marketPlace).adapters(u, m, p);
+
+        // Fetch the principal token for this lend call
+        address pt = IMarketPlace(marketPlace).markets(u, m, p);
+
+        // Get the starting balance for the Lender
+        uint256 starting = IERC20(pt).balanceOf(address(this));
+
+        // Conduct the lend operation to acquire principal tokens
+        (bool success, ) = adapter.delegatecall(
+            abi.encodeWithSignature('lend(bytes calldata inputdata)', d) // TODO: create lend signature
+        );
+
+        if (!success) {
```

Comment:
I actually dont 100% know how delegatecall returns work, but I assume we able to return data other than success as well?

If so I might consider a design change that doesnt require the user to pay for the transfer of PTs back to the lender? 

Mainly I'm trying to think of ways to minimize the number of transfers or at least keep them the same as our current infrastructure.

Reply:
So with delegatecall, there is no need to transfer tokens into or out of the contract -- all of the execution occurs within the original caller's memory space. In this case, we wouldn't be doing any additional transfers. Instead, we'd just be loading in the respective protocols swap logic and executing it within the Lender contract.